File: Interpolator.py
from matplotlib import pyplot as plt
import numpy as np
 
# naive method
def lagrange_bad(x,X,Y):
  n = min(len(X),len(Y))
  m = len(x)  
  Yh = np.zeros(m)
  phi = np.zeros(n)  
  for i in range(m):
    for j in range(n):
      phi[j] = 1.0
      for k in range(n):
        if j != k:
          phi[j] = phi[j]*(x[i]-X[k])/(X[j]-X[k])
      Yh[i] += Y[j] * phi[j]
  return Yh

# better method
def lagrange(x,X,Y):
  n = min(len(X),len(Y)) # same length normally
  phi = np.ones((n,len(x))) 
  for i in range(n):
    for j in range(n):
      if j!=i:
        phi[i,:] *= (x-X[j])/(X[i]-X[j])
      
  return Y @ phi # (1, N) X (N, x) = (1, x)


degP = 12
interA = -1
interB = 1
precision = 2000
f = lambda x : 1/(1 + 25*x**2)

x = np.linspace(interA,interB,precision)
X = np.linspace(interA,interB,degP)
Xtcheb = np.array([np.cos((2*i-1)*np.pi/(2*degP)) for i in range(1,degP+1)]) # set on [-1,1]
Xtcheb = (Xtcheb+1)/2*(interB-interA)+interA  # set on [0,1] then on [-a,b]

Yeq = f(X)
Ytcheb = f(Xtcheb)

plt.figure('Interpolation polynomiale')
plt.scatter(X,Yeq,label='Abscisses équidistantes')
plt.scatter(Xtcheb,Ytcheb,label='Abscisses de Tchebychev')

plt.plot(x, f(x), label='f(x)')
plt.plot(x, lagrange(x, X, Yeq), label="Lagrange (Tchebychev)")
plt.plot(x, lagrange(x, Xtcheb, Ytcheb), label="Lagrange (équidistants)")

plt.legend()
plt.show()
 

File: Interpolator.jl
using Plots
using LaTeXStrings

function lagrange(X, Y, x)
    n = length(X) # or Y, it must have the same shape !
    phi = ones(n, length(x)) # coeff
    for i in 1:n
        for j in 1:n
            if i != j
                phi[i,:] .*= (x .- X[j]) ./ (X[i] .- X[j])
            end
        end
    end
    return transpose( transpose(Array{Float64}(Y)) * Array{Float64}(phi) ) # Y seems have a wrong shape
end

f = x -> 1/(1+25*x^2)
a,b = -1,1
step = 0.01 # pas
x = a:step:b
degP = 15

X = [i/degP for i in 1:degP] # to set on [0,1]
X = (b-a)*X .+ a  # to set on [a,b]
Y = [i for i in f.(X)]
y = lagrange(X, Y, x)

plot(x, y, label="Lagrange (équidistants)", legend=:top)
scatter!(X, Y, label="Abscisses équidistantes")

XT = [cos((2*i-1)*pi/(2*degP)) for i in 1:degP] # set on [-1,1]
XT = (XT.+1)/2*(b-a) .+a  # set on [0,1] then on [-a,b]
YT = [i for i in f.(XT)]
yT = lagrange(XT, YT, x)

plot!(x, yT, label="Lagrange (Tchebychev)")
scatter!(XT, YT, label="Abscisses Tchebychev")

plot!(f, label=L"$f(x) = \frac{1}{1+25x^2}$")
# savefig("Interpolateur_Equi_Tchebychev.png")